### PR TITLE
bears/css: Add PostCSSBear

### PR DESCRIPTION
**Pull request by mr-karan**
This bear uses PostCSS CLI and plugins are
to be defined in .coafile by the user.
The default plugin is `autoprefixer`

Fixes https://github.com/coala-analyzer/coala-bears/issues/282

### CODE REVIEW

File: bears/css/PostCSSBear.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        if postcss_plugins:
+            for plugin in postcss_plugins:
+                self.prerequisite_command = ['postcss', '--use ', plugin]
+                self.prerequisite_fail_msg = "Plugin {} not installed".format(
+                    plugin)
+                self.arguments += (" --use "+plugin)
+        else:
+            # PostCSS needs atleast one plugin to run.
+            # Autoprefixer is the default plugin in that case.
+            self.arguments = "--use autoprefixer"
```
Comment (sils1297) [inline bears/css/PostCSSBear.py]: just error out in that case, not?

File: bears/css/PostCSSBear.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+from coalib.bears.LocalBear import LocalBear
+from coalib.settings.Setting import typed_list
+
+
+class PostCSSBear(Lint, LocalBear):
+    executable = "postcss"
+    diff_message = "PostCSS plugins in action"
+    gives_corrected = True
+
+    def run(self, filename, file, postcss_plugins: typed_list(str)=[]):
```
Comment (sils1297) [inline bears/css/PostCSSBear.py]: don't give a default so coala will ask the user for it, you should document the param a bit better also

File: bears/css/PostCSSBear.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        '''
+        This bear adds ``PostCSS`` utilities to analyze CSS
+        code.
```
Comment (sils1297) [inline bears/css/PostCSSBear.py]: missing newwline

File: bears/css/PostCSSBear.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+from coalib.bearlib.abstractions.Lint import Lint
+from coalib.bears.LocalBear import LocalBear
+from coalib.settings.Setting import typed_list
+
+
+class PostCSSBear(Lint, LocalBear):
+    executable = "postcss"
+    diff_message = "PostCSS plugins in action."
```
Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: Really ? No better possible message ?

File: bears/css/PostCSSBear.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    gives_corrected = True
+
+    def run(self, filename, file, postcss_plugins: typed_list(str)=[]):
+        '''
+        This bear adds ``PostCSS`` utilities to analyze CSS code.
+
+        :param postcss_plugins:     List of PostCSS plugins to use with
```
Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: Unnecessary whitespace after param

File: bears/css/PostCSSBear.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+                                    this bear which can be defined in
+                                    .coafile.
+        '''
+        if postcss_plugins:
```
Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: Would be nicer to do `if not postcss_plugin` and then you don't need an else at all.
makes it easier to read

File: bears/css/PostCSSBear.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+            for plugin in postcss_plugins:
+                self.prerequisite_command = ['postcss', '--use ', plugin]
+                self.prerequisite_fail_msg = "Plugin {} not installed".format(
+                    plugin)
+                self.arguments += (" --use "+plugin)
```
Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: space for `+`
unnecessary ()

File: bears/css/PostCSSBear.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        '''
+        This bear adds ``PostCSS`` utilities to analyze CSS code.
+
+        :param postcss_plugins:     List of PostCSS plugins to use with
+                                    this bear which can be defined in
+                                    .coafile.
+        '''
+        if postcss_plugins:
+            for plugin in postcss_plugins:
+                self.prerequisite_command = ['postcss', '--use ', plugin]
```
Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: This ... isn't used anywhere.
Hmm. I wonder how we can do this.
@sils1297  ?

File: bears/css/PostCSSBear.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+                self.prerequisite_fail_msg = "Plugin {} not installed".format(
+                    plugin)
+                self.arguments += (" --use "+plugin)
+            self.arguments += " {filename}"
+            return self.lint(filename, file)
+        else:
+            self.err("Please specify at least one plugin name.")
```
Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: Does postcss not do anything on it's own without plugins ?

Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: It basically sounds like coala for css :P

File: bears/css/PostCSSBear.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+                                    this bear which can be defined in
+                                    .coafile.
+        '''
+        if postcss_plugins:
+            for plugin in postcss_plugins:
+                self.prerequisite_command = ['postcss', '--use ', plugin]
+                self.prerequisite_fail_msg = "Plugin {} not installed".format(
+                    plugin)
+                self.arguments += (" --use "+plugin)
+            self.arguments += " {filename}"
```
Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: Just wondering ... does it have stdin ?

File: bears/css/PostCSSBear.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    def run(self, filename, file, postcss_plugins: typed_list(str)=[]):
+        '''
+        This bear adds ``PostCSS`` utilities to analyze CSS code.
+
+        :param postcss_plugins:     List of PostCSS plugins to use with
+                                    this bear which can be defined in
+                                    .coafile.
+        '''
+        if postcss_plugins:
+            for plugin in postcss_plugins:
```
Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: Um, Can't we just "--use" multiple plugins ? Would be much faster

Comment (mr-karan) [inline bears/css/PostCSSBear.py]: I tried that, it only uses the first plugin in the argument.

File: bears/css/PostCSSBear.py (hunk 10 above)
Comment (mr-karan) [inline bears/css/PostCSSBear.py]: Don't think so, not mentioned in the docs.

File: bears/css/PostCSSBear.py (hunk 9 above)
Comment (mr-karan) [inline bears/css/PostCSSBear.py]: > The tool itself is a Node.js module that parses CSS into an abstract syntax tree (AST); passes that AST** through any number of “plugin” functions;** and then converts that AST back into a string, which you can output to a file. Each function the AST passes through may or may not transform it; sourcemaps will be generated to keep track of any changes.
The AST provides a straightforward API that developers can use to write plugins

http://davidtheclark.com/its-time-for-everyone-to-learn-about-postcss/

So yeah, without plugins I don't think it can do anything. Much like coala-bears doesn't do anything without specifying a bear.

postcss

File: bears/css/PostCSSBear.py (hunk 11 above)
Comment (mr-karan) [inline bears/css/PostCSSBear.py]: The format is `--use plugin_1 --use plugin_2`

Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: Ahh, nice. Let's use that then ?
Also, what happens if I give `--use plugin1 --use non-existent-plugin` ?
if there's an error message can we show that ?

Comment (mr-karan) [inline bears/css/PostCSSBear.py]: ` self.arguments += " --use " + plugin` Is in a `for` loop, so we are already formatting as required. 

For the second part, yes.
`Error: Cannot find module 'inavlidplugin'`

File: bears/css/PostCSSBear.py (hunk 4 above)
Comment (sils1297) [inline bears/css/PostCSSBear.py]: yes the message needs to be better, maybe we can make some custom ones for the different plugins?

File: bears/css/PostCSSBear.py (hunk 8 above)
Comment (sils1297) [inline bears/css/PostCSSBear.py]: we can't since we allow the user to use arbitrary plugins, we have to do prerequisites checking at runtime if we wanted to d that

File: bears/css/PostCSSBear.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
+from coalib.bearlib.abstractions.Lint import Lint
+from coalib.bears.LocalBear import LocalBear
+from coalib.settings.Setting import typed_list
+
+
+class PostCSSBear(Lint, LocalBear):
+    executable = "postcss"
+    gives_corrected = True
+
+    def run(self, filename, file, postcss_plugins: typed_list(str)=[]):
```
Comment (sils1297) [inline bears/css/PostCSSBear.py]: why do you give an empty list as default and then error out? Just give no default so the user will be asked

File: bears/css/PostCSSBear.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+        :param postcss_plugins: List of PostCSS plugins to use with
+                                this bear which can be defined in
+                                .coafile.
+        '''
+        if not postcss_plugins:
+            self.err("Please specify at least one plugin name.")
+            return []
+
+        for plugin in postcss_plugins:
```
Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: You need to do more analysis on the invalidplugins.
If an invalid plugin is given then you mentioned postcss gives an error message.
 - Is that error message in stdout or stderr ?
 - Is the "corrected file" in stdout or stderr ?
We need to catch that and make Lint [_print_errors()](https://github.com/coala-analyzer/coala/blob/master/coalib/bearlib/abstractions/Lint.py#L102) with it

File: bears/css/PostCSSBear.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+            self.arguments += " --use " + plugin
+            self.arguments += " {filename}"
```
Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: Do we need to `+=` filename for every `--use` ?

File: bears/css/PostCSSBear.py [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+            self.diff_message = "PostCSS plugin {} is analyzing.".format(
```
Comment (AbdealiJK) [inline bears/css/PostCSSBear.py]: This really doesn't work like this !!!!
The last diff_message will be used. The intermediate ones are useless.

**Comment by sils1297** [comment]
@mr-karan ping? What's the state?

**Comment by AbdealiJK** [comment]
fc12178 needs work

**Comment by Makman2** [comment]
https://github.com/coala-analyzer/coala-bears/pull/809 takes over

**Comment by Makman2** [comment]
@mr-karan if you still want to work on that bear, ping me and you can take my branch :)
